### PR TITLE
HSEARCH-4099 Allow option for setting Elasticsearch connection information via full URIs

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-4099

### CODE REVIEW

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 1]
```diff
@@ -105,6 +105,28 @@ The default for this property is `http`.
 
 This property may be set to either `http` or `https`.
 
+==== Provide complete URIs
+
+Alternatively to `hosts` and `protocol`, it is possible to define both the terms using the same property:
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:110]:
```suggestion
 Alternatively, it is possible to define both the protocol and hosts as one or more URIs using a single property:
```

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 2]
```diff
@@ -105,6 +105,28 @@ The default for this property is `http`.
...
+hibernate.search.backend.uris = http://localhost:9200
+----
+
+This property may be set to a String representing an URI such as `http://localhost` or `https://es.mycompany.com:4400`,
+or a String containing multiple such URI strings separated by commas,
+or a `Collection<String>` containing such URI strings.
+
+[WARNING]
+====
+Note that there atrre some constraints we need to respect in order to use this property:
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:123]:
```suggestion
There are some constraints regarding the use of this property:
```

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 3]
```diff
@@ -105,6 +105,28 @@ The default for this property is `http`.
...
+
+* All the uris must have the same protocol.
+* Cannot be used if `hosts` or `protocol` are set.
+* It must contain at least a valid URI.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:127]:
```suggestion
* The provided list of URIs must not be empty.
```

**Review by yrodiere — COMMENTED** [review]
Nice, here are a few comments.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/ServerUris.java [hunk 4]
```diff
@@ -27,7 +29,51 @@ private ServerUris(HttpHost[] hosts, boolean sslEnabled) {
...
+
+		if ( hostAndPortStrings.isPresent() ) {
+			throw log.uriAndHosts( uris.get(), hostAndPortStrings.get() );
+		}
+
+		return fromStrings( uris.get() );
+	}
+
+	private static ServerUris fromStrings(List<String> serverUrisStrings) {
+		if ( serverUrisStrings.isEmpty() ) {
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/ServerUris.java:54]: I don't think it either, I'm adding the test...

[the remark this answers is not in the document]
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/ServerUris.java:54]: At the moment we get an error from the ES client.
``` bash
java.lang.AssertionError: 
Expecting:
  <java.lang.IllegalArgumentException: hosts must not be null nor empty>
to be an instance of:
  <org.hibernate.search.util.common.SearchException>
but was:
  <"java.lang.IllegalArgumentException: hosts must not be null nor empty

...
```

**Comment by fax4ever** [comment]
Thanks for review.
I might have applied the suggestions.

**Comment by yrodiere** [comment]
Merged, thanks

**Pull request merged**